### ISSUE TITLE
Integer key types for `Map`

### ISSUE THREAD

**Issue by uint** [CosmWasm/storey#48]
Also update user docs with storey examples:
- [ ] [IBC Async Ack](https://cosmwasm-docs.vercel.app/ibc/diy-protocol/packet-lifecycle#async-acknowledgement)

**Comment by uint** [comment]
So this got me thinking about bounded iteration. A redesign of map keys might be in order so that we distinguish between dynamically and statically sized keys (`u32` is statically sized, `String` is dynamically sized).

This is important since length-prefixing messes with the order of iteration, but no length-prefixing means bounded iteration will work as expected. I'd like to enable bounded iteration, but only in situations when it's not surprising (e.g. when using statically sized keys).

I also need to move length-prefixing logic from `Map` to the key itself, so that stuff like `u32` can opt out.

tl;dr a bit of a rabbit hole here, I want to make sure we're nudging the design in the right direction when we do this

**Comment by uint** [comment]
[Every Day We Stray Further From The Light](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=292e1c63edce44089ddd53099a780ac7)
